### PR TITLE
`Feature`: Add link previews to chat

### PR DESCRIPTION
**Pull request by julian-wls**
### Problem Description

Sometimes, when all requirements are fulfilled the web app shows a link below a post containing a link. This is currently missing on the Android App.

### Changes

This PR adds link previews to the chat.
The MetisService has been extended to make an additional API call to request a LinkPreview.
Functions have been added to the ConversationViewModel to find links in a post and fetch LinkPreviews accordingly.
Authors can remove link previews, if more than one link preview is available the preview images will be removed (like in the web app).

### Steps for testing

1. Go to any chat.
2. Send links (guaranteed working: stackoverflow.com, ogp.me, php.org)
3. Make sure the preview is shown properly (no preview image if more than one is shown)
4. Remove the preview and make sure the text gets updated and the preview is never shown again.
5. Test step 1-4 for answer messages.

### Screenshots

link_preview2

### CODE REVIEW

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    fun findLinks(text: String): List<Link> {
+        val linkableItems = mutableListOf<Link>()
+
+        val matches = urlRegex.findAll(text)
+        for (match in matches) {
+            val url = match.value
+            val start = match.range.first
+            val end = match.range.last + 1
+
+            val isRemoved = text.getOrNull(start - 1) == '<' && text.getOrNull(end) == '>'
```
Comment (FelberMartin) [inline feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt:17]: rename to `isPreviewRemoved` to avoid confusion

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            val linkableItem = Link(
+                value = url,
+                isLink = true,
+                start = start,
+                end = end,
+                isLinkPreviewRemoved = isRemoved
+            )
+
+            if (!isRemoved) {
+                linkableItems.add(linkableItem)
```
Comment (FelberMartin) [inline feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt:28]: I'm not sure whether I understand correctly, we only add links to the result for which the preview is not removed? Based on the function name, I would not expect this behaviour

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt (hunk 1 above)
Comment (FelberMartin) [inline feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt:17]: We could introduce constants for the "<" and ">" to indicate their purpose

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        return linkableItems
+    }
+
+    fun removeLinkPreview(text: String, urlToSearchFor: String): String {
+        var modifiedContent = text
+        val matches = urlRegex.findAll(modifiedContent).toList()
+
+        for (match in matches) {
+            val url = match.value
+            val normalizedUrl = if (!url.endsWith("/")) urlToSearchFor.trimEnd('/') else urlToSearchFor
```
Comment (FelberMartin) [inline feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt:40]: I think we don't need the if statement here

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            val start = match.range.first
+            val end = match.range.last + 1
+
+            if (url == normalizedUrl || normalizedUrl.contains(url)) {
```
Comment (FelberMartin) [inline feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt:44]: Can you explain why we need this if statement here? 🤔

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt [hunk 5]
```diff
@@ -0,0 +1,51 @@
+package de.tum.informatics.www1.artemis.native_app.feature.metis.conversation.ui.post.util
+
+import de.tum.informatics.www1.artemis.native_app.feature.metis.conversation.service.model.Link
+
+object LinkifyService {
```
Comment (FelberMartin) [inline feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt:5]: I'd argue to choose a different name here to avoid confusion with the other Server-API related "Service" classes we have in the project

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/ConversationViewModel.kt [hunk 6]
```diff
@@ -541,6 +546,58 @@ internal open class ConversationViewModel(
...
+    }
+
+    fun generateLinkPreviews(postContent: String): StateFlow<List<LinkPreview>> =
+        combine(
+            accountService.authToken,
+            serverConfigurationService.serverUrl
+        ) { authToken, serverUrl ->
+            val links = LinkifyService.findLinks(postContent)
+                .filter { it.isLinkPreviewRemoved != true }
+                .take(6)
```
Comment (FelberMartin) [inline feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/ConversationViewModel.kt:571]: No magicNumbers, why 6?

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/ConversationViewModel.kt [hunk 7]
```diff
@@ -541,6 +546,58 @@ internal open class ConversationViewModel(
...
+        serverUrl: String
+    ): LinkPreview? {
+        return retryOnInternet(networkStatusProvider.currentNetworkStatus) {
+            metisService.fetchLinkPreview(
+                url = link.value,
+                authToken = authToken,
+                serverUrl = serverUrl
+            ).bind { preview ->
+                preview?.takeIf {
+                    it.url.isNotEmpty() && it.title.isNotEmpty() && it.description.isNotEmpty() && it.image.isNotEmpty()
```
Comment (FelberMartin) [inline feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/ConversationViewModel.kt:593]: Maybe extract this to a function in LinkPreview? This would also make the purpose of this takeIf more explicit

**Review by FelberMartin — CHANGES_REQUESTED** [review]
Tested the functionality, works as described :) 

It's a little bit frustrating that the link previews are not generated consistently (eg php.org did not work for me on TS2), but there's nothing we can do about that from the Android side.

Besides the code comments, adding unit tests for `LinkifyService` would be awesome :D

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt (hunk 3 above)
Comment (julian-wls) [reply feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt:40]: I think we do, because if url does end with `"/"` but we trim urlToSearchFor to remove the `"/"`, `url == normalizedUrl` will be false even if the link is the same.

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkPreviewUtil.kt [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+            if (url == normalizedUrl || normalizedUrl.contains(url)) {
+                modifiedContent = modifiedContent.removeLinkPreviewTags(start, end, url)
+            }
+        }
+
+        return modifiedContent
+    }
+
+    private fun String.removeLinkPreviewTags(start: Int, end: Int, url: String): String =
```
Comment (FelberMartin) [inline feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkPreviewUtil.kt:60]: Renaming suggestion: `addRemovedLinkPreviewTags`. `removeLinkPreviewTags` suggests to me that the string will somehow be truncated

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt (hunk 4 above)
Comment (FelberMartin) [reply feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt]: Okay I am not sure whether I fully understand, but as long as this scenario/edge case is also covered with tests, it is fine to go from my side :)

File: feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt (hunk 3 above)
Comment (FelberMartin) [reply feature/metis/conversation/src/main/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/util/LinkifyService.kt]: In case we do not have one yet, could we add a unit test for this? :D

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/LinkPreviewTest.kt [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+        assert(result.isEmpty())
+    }
+
+    @Test
+    fun `test GIVEN text with a link WHEN removing link preview THEN it correctly removes the preview`() {
+        val text = "This link https://example.com/something should be removed."
+        val modifiedText = LinkPreviewUtil.removeLinkPreview(text, "https://example.com")
+
+        assertFalse(modifiedText == "<https://example.com/something>")
```
Comment (FelberMartin) [inline feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/post/LinkPreviewTest.kt:43]: This assertion does not align with the test case name, right?

**Review by FelberMartin — CHANGES_REQUESTED** [review]
